p2p/ipld/read.go

```diff
@@ -3,16 +3,215 @@ package ipld
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 
 	"github.com/ipfs/go-cid"
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
 	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
 )
 
-// /////////////////////////////////////
+// //////////////////////////////////////
+//	Retrieve Block Data
+// ////////////////////////////////////
+
+const baseErrorMsg = "failure to retrieve block data:"
+
+var ErrEncounteredTooManyErrors = fmt.Errorf("%s %s", baseErrorMsg, "encountered too many errors")
+var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
+
+// RetrieveBlockData asynchronously fetches block data until the underlying extended
+// data square can be restored and returned.
+func RetrieveBlockData(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	api coreiface.CoreAPI,
+	codec rsmt2d.CodecType,
+) (types.Data, error) {
+	edsWidth := uint32(len(dah.ColumnRoots))
+	originalSquareWidth := edsWidth / 2
+
+	// keep track of leaves using thread safe counter
+	sc := newshareCounter(ctx, edsWidth)
+
+	// convert the row and col roots into Cids
+	rowRoots := dah.RowsRoots.Bytes()
+	rowCids, err := rootsToCids(rowRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	colRoots := dah.ColumnRoots.Bytes()
+	colCids, err := rootsToCids(colRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	// async attempt to fetch each leaf
+	for axisIdx := uint32(0); axisIdx < uint32(len(rowCids)); axisIdx++ {
+		for idx := uint32(0); idx < edsWidth; idx++ {
+			go sc.retrieveShare(colCids[axisIdx], false, axisIdx, idx, api)
+			go sc.retrieveShare(rowCids[axisIdx], true, axisIdx, idx, api)
```

Comment:
But independent from the optimization: what is the reason for trying to request both, rows _and_ columns?
If we care only about the original data, shouldn't one of them be enough? 

Also, I think, we don't need to download up to `edsWidth`. Up to the actual square width should be enough. The rest of the (extended) square can then be recomputed and the erasure coding can then be verified against the DA header.

cc @adlerjohn does this make sense to you too?

Reply:
> what is the reason for trying to request both, rows and columns?

This was a speed optimization. Just whichever comes first.

Attempting to download the entire eds width was also a speed optimization, instead of waiting for the entire original square size to finish, try to download everything but stop once enough has been downloaded to repair.

Reply:
I'm also all for decreasing the number of goroutines, as that is what is causing the test to fail. Apparently we're not supposed to have 8192 ongoing goroutines at once, which we definitely exceed here

Reply:
> Attempting to download the entire eds width was also a speed optimization, instead of waiting for the entire original square size to finish, try to download everything but stop once enough has been downloaded to repair.

Gotcha! That makes sense. It wasn't immediately obvious to me by reading the code. Maybe add a comment on top of the loop stating that this will stop as soon as enough data was downloaded (and how much "enough" means).